src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/PackageReferenceProject.cs

```diff
@@ -135,42 +137,59 @@ public virtual async Task<ProjectPackages> GetInstalledAndTransitivePackagesAsyn
                 else
                 {
                     // Don't mutate cache for threadsafety, instead we works on copy then replace cache when done.
-                    lock (_lockObj)
+                    lock (_installedAndTransitivePackagesLock)
                     {
                         (installedPackages, transitivePackages) = GetInstalledAndTransitivePackagesCacheCopy();
                     }
                 }
             }
 
-            var frameworkSorter = new NuGetFrameworkSorter();
-
             // get installed packages
             List<PackageReference> calculatedInstalledPackages = packageSpec
                 .TargetFrameworks
                 .SelectMany(f => ResolvedInstalledPackagesList(f.Dependencies, f.FrameworkName, targetsList, installedPackages))
                 .GroupBy(p => p.PackageIdentity)
-                .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First())
+                .Select(g => g.OrderBy(p => p.TargetFramework, _frameworkSorter).First())
                 .ToList();
 
             // get transitive packages
             IEnumerable<PackageReference> calculatedTransitivePackages = packageSpec
                 .TargetFrameworks
                 .SelectMany(f => ResolvedTransitivePackagesList(f.FrameworkName, targetsList, installedPackages, transitivePackages))
                 .GroupBy(p => p.PackageIdentity)
-                .Select(g => g.OrderBy(p => p.TargetFramework, frameworkSorter).First());
+                .Select(g => g.OrderBy(p => p.TargetFramework, _frameworkSorter).First());
 
             CounterfactualLoggers.TransitiveDependencies.EmitIfNeeded(); // Emit only one event per VS session
             IEnumerable<TransitivePackageReference> transitivePackagesWithOrigins;
             if (await ExperimentUtility.IsTransitiveOriginExpEnabled.GetValueAsync(token))
             {
-                // Get Transitive Origins
+                // Compute Transitive Origins
+                Dictionary<string, TransitiveEntry> transitiveOrigins;
+                if (IsInstalledAndTransitiveComputationNeeded || TransitiveOriginsCache == null)
+                {
+                    transitiveOrigins = calculatedTransitivePackages.Any() ? ComputeTransitivePackageOrigins(calculatedInstalledPackages, targetsList, token) : new Dictionary<string, TransitiveEntry>();
+                }
+                else
+                {
+                    lock (_transitiveOriginsLock)
+                    {
+                        // Make a copy of the cache to prevent concurrency issues.
+                        transitiveOrigins = new Dictionary<string, TransitiveEntry>(TransitiveOriginsCache);
+                    }
```

Comment:
If `TransitiveOriginsCache` is saved as a local variable, then used in the constructor argument, then it should be thread-safe without needing a lock.

Reply:
Chatted offline. We lock so the cache isn't mutated while we're making a copy of it.